File: world.py
"""
World Grid for EvoSim.

The world is a fixed-size 2-D grid. Each cell can hold at most one
creature. The world also exposes helper methods that creatures use
for sensing their environment.
"""

import numpy as np
from genome import genome_similarity
from config import WORLD_WIDTH, WORLD_HEIGHT


class World:
    """
    Manages the spatial grid and all creature movement.
    """

    def __init__(self, width: int = WORLD_WIDTH, height: int = WORLD_HEIGHT,
                 seed: int = None):
        self.width  = width
        self.height = height
        self.rng    = np.random.default_rng(seed)
        # grid[y][x] = Creature or None
        self._grid  = [[None] * width for _ in range(height)]
        self.creatures = []      # all living creatures this generation
        self.murder_count = 0    # reset each generation

    # ──────────────────────────────────────────────────────────────────────────
    # Placement helpers
    # ──────────────────────────────────────────────────────────────────────────

    def place_creature(self, creature) -> bool:
        """Put a creature at its (x,y) if the cell is free."""
        if self._in_bounds(creature.x, creature.y):
            if self._grid[creature.y][creature.x] is None:
                self._grid[creature.y][creature.x] = creature
                return True
        return False

    def remove_creature(self, creature):
        """Remove creature from grid."""
        if self._in_bounds(creature.x, creature.y):
            if self._grid[creature.y][creature.x] is creature:
                self._grid[creature.y][creature.x] = None

    def clear(self):
        """Remove all creatures from the grid."""
        self._grid = [[None] * self.width for _ in range(self.height)]
        self.creatures = []
        self.murder_count = 0

    def populate(self, creatures: list):
        """
        Place a list of creatures at random empty cells.
        Assigns (x,y) positions before placement.
        """
        self.clear()
        # Generate a shuffled list of all positions
        positions = [(x, y)
                     for y in range(self.height)
                     for x in range(self.width)]
        positions = [positions[i] for i in self.rng.permutation(len(positions))]
        placed = 0
        for creature in creatures:
            if placed >= len(positions):
                break
            creature.x, creature.y = positions[placed]
            self._grid[creature.y][creature.x] = creature
            placed += 1
        self.creatures = list(creatures[:placed])

    # ──────────────────────────────────────────────────────────────────────────
    # Movement
    # ──────────────────────────────────────────────────────────────────────────

    def move_creature(self, creature, dx: int, dy: int) -> bool:
        """
        Attempt to move creature by (dx, dy).
        Returns True on success, False if blocked or OOB.
        Movement is clamped at world boundaries (walls).
        """
        nx = max(0, min(self.width  - 1, creature.x + dx))
        ny = max(0, min(self.height - 1, creature.y + dy))
        if nx == creature.x and ny == creature.y:
            return False     # wall block
        if self._grid[ny][nx] is not None:
            return False     # occupied
        # Move
        self._grid[creature.y][creature.x] = None
        creature.x, creature.y = nx, ny
        self._grid[ny][nx] = creature
        # Update last direction
        if dx != 0 or dy != 0:
            from creature import DIRS
            for i, (ddx, ddy) in enumerate(DIRS):
                if ddx == (1 if dx > 0 else (-1 if dx < 0 else 0)) and \
                   ddy == (1 if dy > 0 else (-1 if dy < 0 else 0)):
                    creature.last_dir = i
                    break
        return True

    def kill_creature_at(self, x: int, y: int):
        """Kill creature at grid cell (x,y) if present."""
        if not self._in_bounds(x, y):
            return
        victim = self._grid[y][x]
        if victim and victim.alive:
            victim.alive = False
            self._grid[y][x] = None
            self.murder_count += 1

    # ──────────────────────────────────────────────────────────────────────────
    # Sensing helpers (used by creatures)
    # ──────────────────────────────────────────────────────────────────────────

    def get_creature(self, x: int, y: int):
        """Return creature at (x,y) or None."""
        if self._in_bounds(x, y):
            return self._grid[y][x]
        return None

    def local_density(self, cx: int, cy: int, radius: int = 2) -> int:
        """Count living creatures within a square of given radius (excl. self)."""
        count = 0
        for dy in range(-radius, radius + 1):
            for dx in range(-radius, radius + 1):
                if dx == 0 and dy == 0:
                    continue
                nx, ny = cx + dx, cy + dy
                if self._in_bounds(nx, ny) and self._grid[ny][nx] is not None:
                    count += 1
        return count

    def forward_density_gradient(self, cx: int, cy: int, dir_idx: int) -> float:
        """
        Measure population density in a cone of 5 cells ahead minus 5 behind.
        Returns a value roughly in [−1, 1].
        """
        from creature import DIRS
        dx, dy = DIRS[dir_idx]
        fwd = sum(
            1 for step in range(1, 6)
            if self._in_bounds(cx + dx*step, cy + dy*step)
            and self._grid[cy + dy*step][cx + dx*step] is not None
        )
        bwd = sum(
            1 for step in range(1, 6)
            if self._in_bounds(cx - dx*step, cy - dy*step)
            and self._grid[cy - dy*step][cx - dx*step] is not None
        )
        return (fwd - bwd) / 5.0

    def genetic_sim_to(self, creature, other) -> float:
        """Genetic similarity between creature and other (0 if no other)."""
        if other is None:
            return 0.0
        return genome_similarity(creature.genome, other.genome)

    # ──────────────────────────────────────────────────────────────────────────
    # Snapshot for visualisation
    # ──────────────────────────────────────────────────────────────────────────

    def snapshot(self):
        """
        Returns two arrays for visualisation:
          positions: list of (x, y) for every living creature
          colors:    list of (r, g, b) tuples
        """
        positions = []
        colors    = []
        for c in self.creatures:
            if c.alive:
                positions.append((c.x, c.y))
                colors.append(c.color)
        return positions, colors

    # ──────────────────────────────────────────────────────────────────────────

    def _in_bounds(self, x: int, y: int) -> bool:
        return 0 <= x < self.width and 0 <= y < self.height

File: dashboard-v2/src/engine/world.js
/* ============================================================
   WORLD MODULE
   ============================================================ */

import { seededRng, shuffle } from './utils';
import { genomeSimilarity } from './genome';
import { NUM_SENSORS, NUM_ACTIONS, DIRS } from './neuralNet';

export class World {
    constructor(W, H) {
        this.W = W; this.H = H;
        this.grid = new Array(W * H).fill(null);
        this.creatures = [];
        this.murders = 0;
        this.rng = seededRng(Date.now());
    }
    idx(x, y) { return y * this.W + x; }
    inBounds(x, y) { return x >= 0 && x < this.W && y >= 0 && y < this.H; }
    get(x, y) { return this.inBounds(x, y) ? this.grid[this.idx(x, y)] : null; }
    set(x, y, c) { if (this.inBounds(x, y)) this.grid[this.idx(x, y)] = c; }
    clear() { this.grid.fill(null); this.creatures = []; this.murders = 0; }

    populate(creatures) {
        this.clear();
        const positions = shuffle(Array.from({ length: this.W * this.H }, (_, i) => i), this.rng);
        for (let i = 0; i < Math.min(creatures.length, positions.length); i++) {
            const pos = positions[i];
            const x = pos % this.W, y = Math.floor(pos / this.W);
            creatures[i].x = x; creatures[i].y = y;
            this.grid[pos] = creatures[i];
        }
        this.creatures = creatures;
    }

    move(creature, dx, dy) {
        const nx = Math.max(0, Math.min(this.W - 1, creature.x + dx));
        const ny = Math.max(0, Math.min(this.H - 1, creature.y + dy));
        if (nx === creature.x && ny === creature.y) return false;
        if (this.get(nx, ny) !== null) return false;
        this.set(creature.x, creature.y, null);
        creature.x = nx; creature.y = ny;
        this.set(nx, ny, creature);
        return true;
    }

    killAt(x, y) {
        const v = this.get(x, y);
        if (v && v.alive) { v.alive = false; this.set(x, y, null); this.murders++; }
    }

    localDensity(cx, cy, r = 2) {
        let count = 0;
        for (let dy = -r; dy <= r; dy++) for (let dx = -r; dx <= r; dx++) {
            if (dx === 0 && dy === 0) continue;
            if (this.inBounds(cx + dx, cy + dy) && this.get(cx + dx, cy + dy)) count++;
        }
        return count;
    }

    stepCreature(creature, simStep, stepsPerGen, killEnabled) {
        if (!creature.alive) return;
        creature.age++;
        creature.oscPhase += (2 * Math.PI) / 30;

        // Sense
        const s = new Float32Array(NUM_SENSORS);
        // Recenter spatial inputs to [-1, 1] to remove East/South bias
        s[0] = (creature.x / (this.W - 1)) * 2 - 1;
        s[1] = (creature.y / (this.H - 1)) * 2 - 1;
        s[2] = creature.age / Math.max(1, stepsPerGen - 1);
        s[3] = this.rng() * 2 - 1; // Recenter random too
        s[4] = (Math.sin(creature.oscPhase) + 1) / 2;
        s[5] = 1 - Math.min(creature.x, this.W - 1 - creature.x) / (this.W / 2);
        s[6] = 1 - Math.min(creature.y, this.H - 1 - creature.y) / (this.H / 2);
        s[7] = Math.min(1, this.localDensity(creature.x, creature.y) / 8);
        const [ddx, ddy] = DIRS[creature.lastDir];
        let fwd = 0, bwd = 0;
        for (let step = 1; step <= 5; step++) {
            if (this.inBounds(creature.x + ddx * step, creature.y + ddy * step) && this.get(creature.x + ddx * step, creature.y + ddy * step)) fwd++;
            if (this.inBounds(creature.x - ddx * step, creature.y - ddy * step) && this.get(creature.x - ddx * step, creature.y - ddy * step)) bwd++;
        }
        s[8] = (fwd - bwd) / 5;
        const ahead = this.get(creature.x + ddx, creature.y + ddy);
        s[9] = ahead ? genomeSimilarity(creature.genome, ahead.genome) : 0;
        // Recenter move inputs to [-1, 1]
        s[10] = DIRS[creature.lastDir][0];
        s[11] = DIRS[creature.lastDir][1];
        s[12] = ahead ? 1 : 0;
        s[13] = 1;

        // Think
        const actions = creature.brain.forward(s);
        let bestAct = -1, bestVal = 0.1;
        for (let i = 0; i < NUM_ACTIONS; i++) {
            if (Math.abs(actions[i]) > bestVal) { bestVal = Math.abs(actions[i]); bestAct = i; }
        }
        if (bestAct < 0) return;

        // Act
        const str = actions[bestAct];
        if (bestAct === 0) { const dx2 = str > 0 ? 1 : -1; if (this.move(creature, dx2, 0)) { creature.lastDir = str > 0 ? 0 : 4; } }
        else if (bestAct === 1) { const dy2 = str > 0 ? 1 : -1; if (this.move(creature, 0, dy2)) { creature.lastDir = str > 0 ? 2 : 6; } }
        else if (bestAct === 2) { const d = Math.floor(this.rng() * 8); if (this.move(creature, DIRS[d][0], DIRS[d][1])) creature.lastDir = d; }
        else if (bestAct === 3) { this.move(creature, DIRS[creature.lastDir][0], DIRS[creature.lastDir][1]); }
        else if (bestAct === 4) { creature.lastDir = (creature.lastDir + 7) % 8; this.move(creature, DIRS[creature.lastDir][0], DIRS[creature.lastDir][1]); }
        else if (bestAct === 5) { creature.lastDir = (creature.lastDir + 1) % 8; this.move(creature, DIRS[creature.lastDir][0], DIRS[creature.lastDir][1]); }
        else if (bestAct === 6) { creature.lastDir = (creature.lastDir + 4) % 8; this.move(creature, DIRS[creature.lastDir][0], DIRS[creature.lastDir][1]); }
        else if (bestAct === 7 && killEnabled) { this.killAt(creature.x + DIRS[creature.lastDir][0], creature.y + DIRS[creature.lastDir][1]); }
    }
}
